### PR TITLE
Add guidance on context and URL caching in general

### PR DESCRIPTION
**Pull request by msporny**
This PR attempts to address issue #170, raised by PING and security review, by adding guidance on context and URL caching, in general. It also adds a security consideration and a privacy consideration on network requests and network request fingerprinting.

/cc @kdenhartog

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/pull/172.html" title="Last updated on Aug 26, 2023, 4:50 PM UTC (6eb0838)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/172/05112f9...6eb0838.html" title="Last updated on Aug 26, 2023, 4:50 PM UTC (6eb0838)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -3223,7 +3281,40 @@ <h3>Selective Disclosure</h3>
...
+over a network connection.
+        </p>
+
+        <p>
+For example, creators of <a>conforming documents</a> might craft unique
+per-document URLs for <a href="#data-model">JSON-LD Contexts</a> and
+<a>verification methods</a>. When verifying such a document, a verifier fetching
+that information from the network would reveal their interest in the
+<a>conforming document</a>, to the creator of the document, which might lead to
+a mismatch in privacy expectations among any entity that is not the creator of
```
Comment (dlongley) [inline index.html]:
```suggestion
a mismatch in privacy expectations for any entity that is not the creator of
```

File: index.html [hunk 2]
```diff
@@ -1868,6 +1880,22 @@ <h2>Contexts and Vocabularies</h2>
...
+equivalent mechanism.
+        </p>
+
+        <p>
+Implementations MAY load application-specific JSON-LD Context files from the
+network during development, but SHOULD permanently cache JSON-LD Context files
+utilized in <a>conforming documents</a> in production settings to increase the
+security and privacy characteristics of solutions. Approaches to caching as
+described above, or functionally equivalent mechanisms, MAY be utilized to
+achieve caching goals.
```
Comment (dlongley) [inline index.html]:
```suggestion
Implementations MAY load application-specific JSON-LD Context files from the
network during development, but SHOULD permanently cache JSON-LD Context files
utilized in <a>conforming documents</a> in production settings to increase the
security and privacy characteristics of solutions. Approaches to caching as
described above, or functionally equivalent mechanisms, MAY be utilized to
achieve caching goals.
        </p>
        <p>
Some applications, such as digital wallets that are capable of holding arbitrary
<a>verifiable credentials</a> (from any issuer, using any contexts) or other
data-integrity-protected documents, might need to be able to load externally
linked resources, such as JSON-LD Context files, in production settings. This is
expected to increase user choice, scalability, and decentralized upgrades in the
ecosystem over time. Authors of these applications are advised to read the
security and privacy sections of this document for further considerations.
        </p>
```

**Review by dlongley — APPROVED** [review]
See comment trying to note the benefit of the use of URLs and acknowledgment that some production applications might need to load them.

**Review by Wind4Greg — APPROVED** [review]
Good privacy advice.

**Review by brianorwhatever — APPROVED** [review]

**Review by kdenhartog — APPROVED** [review]
Text reads fine as is so approving. I'd suggest linking to https://github.com/ietf-wg-ohai/oblivious-http as one method to addressing privacy concerns when needing to resolve in production instances as well. There's others as well such as generic proxying through a trusted third party (e.g. wallets proxy requestthrough their backend services or mapping URL resolutions in wallet JSON-LD document loaders to trusted services where they can be resolved) that can be used as alternatives so it doesn't need to require any particular solution. The main thinking here is that suggesting solutions is more likely to lead to them getting addressed but the current text states this well enough in my opinion that it doesn't need to be blocked on this addition.

File: index.html [hunk 3]
```diff
@@ -1868,6 +1880,32 @@ <h2>Contexts and Vocabularies</h2>
 `curl -sL -H "Accept: &lt;MEDIA_TYPE>" &lt;DOCUMENT_URL> | openssl dgst -&ltDIGEST_ALGORITHM> -binary | openssl base64 -nopad -a`.
         </p>
 
+        <p>
+Application-specific vocabulary and specification authors SHOULD ensure that
+their JSON-LD Context and vocabulary files are permanently cacheable using
+either the approaches to caching as described above, or any other functionally
+equivalent mechanism.
```
Comment (TallTed) [inline index.html]:
```suggestion
Authors of application-specific vocabularies and specifications SHOULD ensure
that their JSON-LD context and vocabulary files are permanently cacheable
using the approaches to caching described above or a functionally equivalent
mechanism.
```

File: index.html [hunk 4]
```diff
@@ -1868,6 +1880,32 @@ <h2>Contexts and Vocabularies</h2>
...
+        </p>
+
+        <p>
+Implementations MAY load application-specific JSON-LD Context files from the
+network during development, but SHOULD permanently cache JSON-LD Context files
+utilized in <a>conforming documents</a> in production settings to increase the
+security and privacy characteristics of solutions. Approaches to caching as
+described above, or functionally equivalent mechanisms, MAY be utilized to
+achieve caching goals.
```
Comment (TallTed) [inline index.html]:
```suggestion
Implementations MAY load application-specific JSON-LD context files from the
network during development, but SHOULD permanently cache JSON-LD context files
used in <a>conforming documents</a> in production settings to increase their
security and privacy characteristics. Caching goals MAY be achieved through
approaches such as those described above or functionally equivalent mechanisms.
```

File: index.html [hunk 5]
```diff
@@ -1868,6 +1880,32 @@ <h2>Contexts and Vocabularies</h2>
...
+        <p>
+Some applications, such as digital wallets that are capable of holding arbitrary
+<a>verifiable credentials</a> (from any issuer, using any contexts) or other
+data-integrity-protected documents, might need to be able to load externally
+linked resources, such as JSON-LD Context files, in production settings. This is
+expected to increase user choice, scalability, and decentralized upgrades in the
+ecosystem over time. Authors of these applications are advised to read the
+security and privacy sections of this document for further considerations.
+        </p>
+        </p>
```
Comment (TallTed) [inline index.html]:
```suggestion
Some applications, such as digital wallets, that are capable of holding arbitrary
<a>verifiable credentials</a> or other data-integrity-protected documents, from
any issuer and using any contexts, might need to be able to load externally
linked resources, such as JSON-LD context files, in production settings. This is
expected to increase user choice, scalability, and decentralized upgrades in the
ecosystem over time. Authors of such applications are advised to read the
security and privacy sections of this document for further considerations.
        </p>
```

File: index.html [hunk 6]
```diff
@@ -1868,6 +1880,32 @@ <h2>Contexts and Vocabularies</h2>
...
+
         <p>
 For further information regarding processing of JSON-LD Contexts and
 Vocabularies, see <a data-cite="?VC-DATA-MODEL-2.0#base-context">Verifiable
```
Comment (TallTed) [inline index.html]: Taking my `contexts` and `vocabularies` lowercasing cue from https://www.w3.org/TR/json-ld11/ where they are only capitalized in section titles.
```suggestion
For further information regarding processing of JSON-LD contexts and
vocabularies, see <a data-cite="?VC-DATA-MODEL-2.0#base-context">Verifiable
```

File: index.html [hunk 7]
```diff
@@ -3128,6 +3166,36 @@ <h3>Canonicalization Method Correctness</h3>
...
+      <section>
+        <h3>Network Requests</h3>
+
+        <p>
+This specification is designed in such a way that no network requests are
+required in order to verify a proof on <a>conforming secured documents</a>.
+Readers might note, however, that <a href="#data-model">JSON-LD Contexts</a>
+and <a>verification methods</a> can contain URLs that might be retrieved
+over a network connection. This concern exists for any URL that might be
+loaded from the network during or after verification.
```
Comment (TallTed) [inline index.html]:
```suggestion
This specification is designed in such a way that no network requests are
required when verifying a proof on a <a>conforming secured document</a>.
Readers might note, however, that <a href="#data-model">JSON-LD contexts</a>
and <a>verification methods</a> can contain URLs that might be retrieved
over a network connection. This concern exists for any URL that might be
loaded from the network during or after verification.
```

File: index.html [hunk 8]
```diff
@@ -3128,6 +3166,36 @@ <h3>Canonicalization Method Correctness</h3>
...
+        </p>
+        <p>
+To the extent possible, implementers are urged to permanently or aggressively
+cache such information in order to reduce the attack surface on an
+implementation that might need to go out to the network and fetch such URLs. For
+example, caching techniques for <a href="#data-model">JSON-LD Contexts</a> are
+described in Section <a href="#contexts-and-vocabularies"></a> and some
+<a>verification methods</a>, such as `did:key` [[?DID-KEY]], do not need to be
+fetched from the network.
```
Comment (TallTed) [inline index.html]:
```suggestion
To the extent possible, implementers are urged to permanently or aggressively
cache such information to reduce the attack surface on an
implementation that might need to fetch such URLs over the network. For
example, caching techniques for <a href="#data-model">JSON-LD contexts</a> are
described in Section <a href="#contexts-and-vocabularies"></a>, and some
<a>verification methods</a>, such as `did:key` [[?DID-KEY]], do not need to be
fetched from the network at all.
```

File: index.html [hunk 9]
```diff
@@ -3128,6 +3166,36 @@ <h3>Canonicalization Method Correctness</h3>
...
+        </p>
+        <p>
+When it is not possible to cache the information, such as when a specific HTTP
```
Comment (TallTed) [inline index.html]:
```suggestion
When it is not possible to use cached information, such as when a specific HTTP
```

File: index.html [hunk 10]
```diff
@@ -3223,7 +3291,40 @@ <h3>Selective Disclosure</h3>
 disclosing information while preserving privacy, see
 Section <a href="#unlinkability"></a>.
         </p>
+      </section>
+
+      <section>
+        <h3>Network Request Fingerprinting</h3>
```
Comment (TallTed) [inline index.html]:
```suggestion
        <h3>Fingerprinting Network Requests</h3>
```

File: index.html [hunk 11]
```diff
@@ -3223,7 +3291,40 @@ <h3>Selective Disclosure</h3>
...
+
+        <p>
+Fingerprinting concerns exists for any URL that might be loaded from the network
+during or after <a href="#verify-proof">proof verification</a>. This
+specification is designed in such a way that no network requests are necessary
+in order to verify a proof on <a>conforming secured documents</a>. Readers might
+note, however, that <a href="#data-model">JSON-LD Contexts</a> and
+<a>verification methods</a> can contain resource URLs that might be retrieved
+over a network connection.
```
Comment (TallTed) [inline index.html]:
```suggestion
Fingerprinting concerns exist for any URL that might be loaded from the network
during or after <a href="#verify-proof">proof verification</a>. This
specification is designed in such a way that no network requests are necessary
when verifying a proof on a <a>conforming secured document</a>. Readers might
note, however, that <a href="#data-model">JSON-LD contexts</a> and
<a>verification methods</a> can contain resource URLs that might be retrieved
over a network connection.
```

File: index.html [hunk 12]
```diff
@@ -3223,7 +3291,40 @@ <h3>Selective Disclosure</h3>
...
+        </p>
+
+        <p>
+For example, creators of <a>conforming documents</a> might craft unique
+per-document URLs for <a href="#data-model">JSON-LD Contexts</a> and
+<a>verification methods</a>. When verifying such a document, a verifier fetching
+that information from the network would reveal their interest in the
+<a>conforming document</a>, to the creator of the document, which might lead to
+a mismatch in privacy expectations for any entity that is not the creator of
+the document.
```
Comment (TallTed) [inline index.html]:
```suggestion
For example, creators of <a>conforming documents</a> might craft unique
per-document URLs for <a href="#data-model">JSON-LD contexts</a> and
<a>verification methods</a>. When verifying such a document, a verifier fetching
that information from the network would reveal their interest in the
<a>conforming document</a> to the creator of the document, which might lead to
a mismatch in privacy expectations for any entity that is not the creator of
the document.
```

File: index.html [hunk 13]
```diff
@@ -3223,7 +3291,40 @@ <h3>Selective Disclosure</h3>
...
+        </p>
 
+        <p>
+Implementers are urged to follow the guidance in Section
+<a href="#network-requests"></a> on URL caching and implementing defensively
+when fetching URLs from the network. Additionally, heuristics might be utilized
+to determine if creators of <a>conforming documents</a> are using fingerprinting
+URLs in a way that might violate privacy expectations. These heuristics could be
+used to display warnings to entities that might process documents containing
+suspected fingerprinting URLs.
```
Comment (TallTed) [inline index.html]:
```suggestion
when fetching URLs from the network. Additionally, heuristics might be used
to determine whether creators of <a>conforming documents</a> are using fingerprinting
URLs in a way that might violate privacy expectations. These heuristics could be
used to display warnings to entities that might process documents containing
suspected fingerprinting URLs.
```

**Review by TallTed — CHANGES_REQUESTED** [review]
A number of grammar and language tweaks.

Note that `context/s` and `vocabulary/ies` are only capitalized in section titles in the JSON-LD spec, and I think that we should do the same. I have not applied this to the whole document, only to the areas touched by this PR. This might call for a distinct PR and/or Issue.

**Comment by msporny** [comment]
@TallTed wrote:
> Note that context/s and vocabulary/ies are only capitalized in section titles in the JSON-LD spec, and I think that we should do the same. 

Agreed, I have raised #183 to track this issue.

@kdenhartog wrote:
> I'd suggest linking to https://github.com/ietf-wg-ohai/oblivious-http as one method to addressing privacy concerns when needing to resolve in production instances as well. 

Added in https://github.com/w3c/vc-data-integrity/pull/172/commits/6eb08386692550e2e1cdfef95ba33bc823db7934.

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**